### PR TITLE
Make the Rust Process own the C Process

### PR DESCRIPTION
**Pull request by sporksmith**

### CODE REVIEW

**Review by stevenengler — APPROVED** [review]

File: src/main/host/process.rs [hunk 1]
```diff
@@ -45,36 +50,50 @@ impl TryFrom<ProcessId> for libc::pid_t {
...
+        // SAFETY: The Process itself is wrapped in a RootedRefCell, which ensures
+        // it can only be accessed by one thread at a time. Whenever we access its cprocess,
+        // we ensure that the pointer doesn't "escape" in a way that would allow it to be
+        // accessed by threads that don't have access to the enclosing Process.
+        let cprocess = unsafe { SyncSendPointer::new(p) };
+        let process = RootedRc::new(root, RootedRefCell::new(root, Process { cprocess }));
+        // We're storing a raw pointer to the inner RootedRefCell here. This is a bit
+        // fragile, but should be ok since its never moved out of the enclosing RootedRc,
+        // so its address won't change. Since the Rust Process owns the C Process, the
+        // Rust Process will outlive the C Process.
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply src/main/host/process.rs:123]: Done

File: src/main/host/syscall/handler/ioctl.rs [hunk 2]
```diff
@@ -42,11 +43,13 @@ impl SyscallHandler {
             CompatFile::New(file) => file,
             // if it's a legacy file, use the C syscall handler instead
             CompatFile::Legacy(_) => {
+                drop(desc_table);
                 return Self::legacy_syscall(c::syscallhandler_ioctl, ctx, args);
             }
         };
 
         let file = file.inner_file().clone();
+        drop(desc_table);
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply src/main/host/syscall/handler/ioctl.rs]: I like this idea, and usually VSCode would make this sort of rename easy, but for some reason in this case it's missing a lot of things. Going to leave this for a follow-up.

File: src/main/host/descriptor/descriptor_table.rs [hunk 3]
```diff
@@ -120,6 +119,30 @@ impl DescriptorTable {
...
+        desc: Descriptor,
+        new_fd: u32,
+    ) -> Option<Descriptor> {
+        self.set(new_fd, desc)
+    }
+
+    /// Deregister the descriptor with the given fd handle and return it.
+    pub fn deregister_descriptor(&mut self, fd: u32) -> Option<Descriptor> {
+        self.remove(fd)
+    }
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply src/main/host/descriptor/descriptor_table.rs]: I think the `register_*` methods are a nicer public interface than `set`. I notice that `remove` is only called from this one place though, so inlined it here.

File: src/main/host/process.c [hunk 4]
```diff
@@ -319,16 +319,17 @@ static void _process_handleProcessExit(Process* proc) {
 }
 
 static void _process_terminate(Process* proc) {
+    if (!process_isRunning(proc)) {
+        trace("Already dead");
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply src/main/host/process.c]: Yes, it should have already been done when the process stopped running by itself. Added some additional checks and comments.

**Pull request merged**